File: server-go/internal/store/json_store.go
package store

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"

	"ttrpg-companion/server-go/internal/model"
)

// JsonCompendiumStore loads compendium data from JSON files at startup.
type JsonCompendiumStore struct {
	spells       []model.Spell
	monsters     []model.Monster
	magicalItems []model.MagicalItem
}

// NewJsonCompendiumStore loads all three compendium JSON files from dataDir.
func NewJsonCompendiumStore(dataDir string) (*JsonCompendiumStore, error) {
	spells, err := loadSpells(filepath.Join(dataDir, "spells.json"))
	if err != nil {
		return nil, fmt.Errorf("loading spells: %w", err)
	}

	monsters, err := loadMonsters(filepath.Join(dataDir, "monsters.json"))
	if err != nil {
		return nil, fmt.Errorf("loading monsters: %w", err)
	}

	items, err := loadMagicalItems(filepath.Join(dataDir, "magical-items.json"))
	if err != nil {
		return nil, fmt.Errorf("loading magical items: %w", err)
	}

	return &JsonCompendiumStore{
		spells:       spells,
		monsters:     monsters,
		magicalItems: items,
	}, nil
}

func (s *JsonCompendiumStore) GetSpells() []model.Spell             { return s.spells }
func (s *JsonCompendiumStore) GetMonsters() []model.Monster         { return s.monsters }
func (s *JsonCompendiumStore) GetMagicalItems() []model.MagicalItem { return s.magicalItems }

// ── Loaders ───────────────────────────────────────────────────────────────────

func loadSpells(path string) ([]model.Spell, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	var raw []model.SpellJson
	if err = json.Unmarshal(data, &raw); err != nil {
		return nil, err
	}

	spells := make([]model.Spell, 0, len(raw))
	for _, r := range raw {
		if s, ok := spellFromJson(r); ok {
			spells = append(spells, s)
		}
	}
	return spells, nil
}

func loadMonsters(path string) ([]model.Monster, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	var raw []model.MonsterJson
	if err = json.Unmarshal(data, &raw); err != nil {
		return nil, err
	}

	monsters := make([]model.Monster, 0, len(raw))
	for _, r := range raw {
		if m, ok := monsterFromJson(r); ok {
			monsters = append(monsters, m)
		}
	}
	return monsters, nil
}

func loadMagicalItems(path string) ([]model.MagicalItem, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	var raw []model.MagicalItemJson
	if err = json.Unmarshal(data, &raw); err != nil {
		return nil, err
	}

	items := make([]model.MagicalItem, 0, len(raw))
	for _, r := range raw {
		if item, ok := magicalItemFromJson(r); ok {
			items = append(items, item)
		}
	}
	return items, nil
}

// ── Converters ────────────────────────────────────────────────────────────────

func spellFromJson(spell model.SpellJson) (model.Spell, bool) {
	if spell.ID == nil || *spell.ID == "" {
		fmt.Println("WARNING: skipping spell with missing id")
		return model.Spell{}, false
	}
	id := *spell.ID
	if spell.Source == nil {
		fmt.Printf("WARNING: skipping spell '%s': missing source\n", id)
		return model.Spell{}, false
	}
	if spell.Level == nil {
		fmt.Printf("WARNING: skipping spell '%s': missing level\n", id)
		return model.Spell{}, false
	}
	if spell.School == nil {
		fmt.Printf("WARNING: skipping spell '%s': missing school\n", id)
		return model.Spell{}, false
	}
	if spell.Concentration == nil {
		fmt.Printf("WARNING: skipping spell '%s': missing concentration\n", id)
		return model.Spell{}, false
	}
	if spell.Ritual == nil {
		fmt.Printf("WARNING: skipping spell '%s': missing ritual\n", id)
		return model.Spell{}, false
	}
	if spell.Components == nil {
		fmt.Printf("WARNING: skipping spell '%s': missing components\n", id)
		return model.Spell{}, false
	}
	if spell.AvailableClasses == nil {
		fmt.Printf("WARNING: skipping spell '%s': missing availableClasses\n", id)
		return model.Spell{}, false
	}
	if len(spell.Translations) == 0 {
		fmt.Printf("WARNING: skipping spell '%s': missing translations\n", id)
		return model.Spell{}, false
	}

	translations := make(map[string]model.SpellTranslation, len(spell.Translations))
	for locale, t := range spell.Translations {
		translations[locale] = model.SpellTranslation{
			Name:                derefString(t.Name),
			CastingTime:         derefString(t.CastingTime),
			Range:               derefString(t.Range),
			Duration:            derefString(t.Duration),
			MaterialDescription: t.MaterialDescription,
			Description:         derefString(t.Description),
		}
	}

	return model.Spell{
		ID:            id,
		Source:        *spell.Source,
		Level:         *spell.Level,
		School:        *spell.School,
		Concentration: *spell.Concentration,
		Ritual:        *spell.Ritual,
		Components: model.SpellComponents{
			Verbal:   spell.Components.Verbal,
			Somatic:  spell.Components.Somatic,
			Material: spell.Components.Material,
		},
		AvailableClasses: spell.AvailableClasses,
		Translations:     translations,
	}, true
}

func monsterFromJson(r model.MonsterJson) (model.Monster, bool) {
	if r.ID == nil || *r.ID == "" {
		fmt.Println("WARNING: skipping monster with missing id")
		return model.Monster{}, false
	}
	id := *r.ID
	if r.Source == nil {
		fmt.Printf("WARNING: skipping monster '%s': missing source\n", id)
		return model.Monster{}, false
	}
	if r.Type == nil {
		fmt.Printf("WARNING: skipping monster '%s': missing type\n", id)
		return model.Monster{}, false
	}
	if r.Size == nil {
		fmt.Printf("WARNING: skipping monster '%s': missing size\n", id)
		return model.Monster{}, false
	}
	if r.Alignment == nil {
		fmt.Printf("WARNING: skipping monster '%s': missing alignment\n", id)
		return model.Monster{}, false
	}
	if r.Abilities == nil {
		fmt.Printf("WARNING: skipping monster '%s': missing abilities\n", id)
		return model.Monster{}, false
	}
	if len(r.Translations) == 0 {
		fmt.Printf("WARNING: skipping monster '%s': missing translations\n", id)
		return model.Monster{}, false
	}

	translations := make(map[string]model.Translation, len(r.Translations))
	for locale, t := range r.Translations {
		if t.Name == nil {
			fmt.Printf("WARNING: monster '%s' locale '%s': missing name\n", id, locale)
			continue
		}
		if t.Description == nil {
			fmt.Printf("WARNING: monster '%s' locale '%s': missing description\n", id, locale)
			continue
		}
		if t.Speed == nil {
			fmt.Printf("WARNING: monster '%s' locale '%s': missing speed\n", id, locale)
			continue
		}
		if t.Senses == nil {
			fmt.Printf("WARNING: monster '%s' locale '%s': missing senses\n", id, locale)
			continue
		}
		translations[locale] = model.Translation{
			Name:        *t.Name,
			Subtype:     t.Subtype,
			Description: *t.Description,
			Speed:       *t.Speed,
			Senses:      *t.Senses,
			Languages:   orEmptySlice(t.Languages),
		}
	}
	if len(translations) == 0 {
		fmt.Printf("WARNING: skipping monster '%s': no valid translations\n", id)
		return model.Monster{}, false
	}

	var cr float32
	if r.ChallengeRating != nil {
		cr = *r.ChallengeRating
	}
	ac := 10
	if r.ArmorClass != nil {
		ac = *r.ArmorClass
	}
	var hp int
	if r.MaxHitPoints != nil {
		hp = *r.MaxHitPoints
	}

	return model.Monster{
		ID:                  id,
		Source:              *r.Source,
		Type:                *r.Type,
		Size:                *r.Size,
		Alignment:           *r.Alignment,
		ChallengeRating:     cr,
		ArmorClass:          ac,
		MaxHitPoints:        hp,
		HitDice:             derefString(r.HitDice),
		Abilities:           abilityGroupFromJson(r.Abilities),
		Skills:              orEmptyStringMap(r.Skills),
		DamageAffinities:    orEmptyStringMap(r.DamageAffinities),
		ConditionImmunities: orEmptyBoolMap(r.ConditionImmunities),
		Translations:        translations,
	}, true
}

func abilityGroupFromJson(a *model.AbilitiesJson) model.Abilities {
	if a == nil {
		return model.Abilities{
			Str: model.Ability{Value: 10},
			Dex: model.Ability{Value: 10},
			Con: model.Ability{Value: 10},
			Int: model.Ability{Value: 10},
			Wis: model.Ability{Value: 10},
			Cha: model.Ability{Value: 10},
		}
	}
	return model.Abilities{
		Str: abilityFromJson(a.Str),
		Dex: abilityFromJson(a.Dex),
		Con: abilityFromJson(a.Con),
		Int: abilityFromJson(a.Int),
		Wis: abilityFromJson(a.Wis),
		Cha: abilityFromJson(a.Cha),
	}
}

func abilityFromJson(a *model.AbilityJson) model.Ability {
	if a == nil {
		return model.Ability{Value: 10}
	}
	v := 10
	if a.Value != nil {
		v = *a.Value
	}
	return model.Ability{Value: v, SavingThrowProficiency: a.SavingThrowProficiency}
}

func magicalItemFromJson(r model.MagicalItemJson) (model.MagicalItem, bool) {
	if r.ID == nil || *r.ID == "" {
		fmt.Println("WARNING: skipping magical item with missing id")
		return model.MagicalItem{}, false
	}
	id := *r.ID
	if r.Source == nil {
		fmt.Printf("WARNING: skipping magical item '%s': missing source\n", id)
		return model.MagicalItem{}, false
	}
	if r.Type == nil {
		fmt.Printf("WARNING: skipping magical item '%s': missing type\n", id)
		return model.MagicalItem{}, false
	}
	if r.Rarity == nil {
		fmt.Printf("WARNING: skipping magical item '%s': missing rarity\n", id)
		return model.MagicalItem{}, false
	}
	if len(r.Translations) == 0 {
		fmt.Printf("WARNING: skipping magical item '%s': missing translations\n", id)
		return model.MagicalItem{}, false
	}

	translations := make(map[string]model.MagicalItemTranslation, len(r.Translations))
	for locale, t := range r.Translations {
		if t.Name == nil {
			fmt.Printf("WARNING: magical item '%s' locale '%s': missing name\n", id, locale)
			continue
		}
		if t.Description == nil {
			fmt.Printf("WARNING: magical item '%s' locale '%s': missing description\n", id, locale)
			continue
		}
		translations[locale] = model.MagicalItemTranslation{
			Name:        *t.Name,
			Subtype:     t.Subtype,
			Description: *t.Description,
		}
	}
	if len(translations) == 0 {
		fmt.Printf("WARNING: skipping magical item '%s': no valid translations\n", id)
		return model.MagicalItem{}, false
	}

	attunement := r.Attunement != nil && *r.Attunement

	return model.MagicalItem{
		ID:           id,
		Source:       *r.Source,
		Type:         *r.Type,
		Rarity:       *r.Rarity,
		Attunement:   attunement,
		Translations: translations,
	}, true
}

// ── Nil-safe helpers ──────────────────────────────────────────────────────────

func derefString(s *string) string {
	if s == nil {
		return ""
	}
	return *s
}

func orEmptySlice(s []string) []string {
	if s == nil {
		return []string{}
	}
	return s
}

func orEmptyStringMap(m map[string]string) map[string]string {
	if m == nil {
		return map[string]string{}
	}
	return m
}

func orEmptyBoolMap(m map[string]bool) map[string]bool {
	if m == nil {
		return map[string]bool{}
	}
	return m
}

File: server-rust/src/store/json_store.rs
use std::collections::HashMap;
use std::path::Path;

use serde::Deserialize;

use crate::models::{
    monster::{Abilities, Ability, Monster, Speeds, Translation},
    magical_item::{MagicalItem, MagicalItemTranslation},
    spell::{Spell, SpellComponents, SpellTranslation},
};
use crate::store::CompendiumStore;

pub struct JsonCompendiumStore {
    spells: Vec<Spell>,
    monsters: Vec<Monster>,
    magical_items: Vec<MagicalItem>,
}

impl JsonCompendiumStore {
    pub fn load(data_dir: &str) -> Result<Self, Box<dyn std::error::Error>> {
        let base = Path::new(data_dir);

        let spells_json = std::fs::read_to_string(base.join("spells.json"))?;
        let monsters_json = std::fs::read_to_string(base.join("monsters.json"))?;
        let items_json = std::fs::read_to_string(base.join("magical-items.json"))?;

        let spells = parse_spells(&spells_json)?;
        let monsters = parse_monsters(&monsters_json)?;
        let magical_items = parse_magical_items(&items_json)?;

        Ok(Self { spells, monsters, magical_items })
    }
}

impl CompendiumStore for JsonCompendiumStore {
    fn get_spells(&self) -> &[Spell] {
        &self.spells
    }

    fn get_monsters(&self) -> &[Monster] {
        &self.monsters
    }

    fn get_magical_items(&self) -> &[MagicalItem] {
        &self.magical_items
    }
}

// ── Raw JSON structs ──────────────────────────────────────────────────────────

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct SpellJson {
    id: Option<String>,
    source: Option<String>,
    level: Option<i32>,
    school: Option<String>,
    concentration: Option<bool>,
    ritual: Option<bool>,
    components: Option<SpellComponentsJson>,
    available_classes: Option<Vec<String>>,
    translations: Option<HashMap<String, SpellTranslation>>,
}

#[derive(Deserialize)]
struct SpellComponentsJson {
    verbal: bool,
    somatic: bool,
    material: bool,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct MonsterJson {
    id: Option<String>,
    source: Option<String>,
    #[serde(rename = "type")]
    monster_type: Option<String>,
    size: Option<String>,
    alignment: Option<String>,
    challenge_rating: Option<f32>,
    armor_class: Option<i32>,
    max_hit_points: Option<i32>,
    hit_dice: Option<String>,
    abilities: Option<AbilitiesJson>,
    speeds: Option<SpeedsJson>,
    #[serde(default)]
    skills: HashMap<String, String>,
    #[serde(default)]
    damage_affinities: HashMap<String, String>,
    #[serde(default)]
    condition_immunities: HashMap<String, bool>,
    translations: Option<HashMap<String, TranslationJson>>,
}

#[derive(Deserialize)]
struct SpeedsJson {
    walk: Option<i32>,
    fly: Option<i32>,
    swim: Option<i32>,
    climb: Option<i32>,
    burrow: Option<i32>,
    hover: Option<bool>,
}

#[derive(Deserialize)]
struct AbilitiesJson {
    str: Option<AbilityJson>,
    dex: Option<AbilityJson>,
    con: Option<AbilityJson>,
    int: Option<AbilityJson>,
    wis: Option<AbilityJson>,
    cha: Option<AbilityJson>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct AbilityJson {
    value: Option<i32>,
    saving_throw_proficiency: Option<String>,
}

#[derive(Deserialize)]
struct TranslationJson {
    name: Option<String>,
    subtype: Option<String>,
    description: Option<String>,
    senses: Option<String>,
    languages: Option<Vec<String>>,
}

#[derive(Deserialize)]
struct MagicalItemJson {
    id: Option<String>,
    source: Option<String>,
    #[serde(rename = "type")]
    item_type: Option<String>,
    rarity: Option<String>,
    attunement: Option<bool>,
    translations: Option<HashMap<String, MagicalItemTranslationJson>>,
}

#[derive(Deserialize)]
struct MagicalItemTranslationJson {
    name: Option<String>,
    subtype: Option<String>,
    description: Option<String>,
}

// ── Converters ────────────────────────────────────────────────────────────────

fn parse_spells(json: &str) -> Result<Vec<Spell>, Box<dyn std::error::Error>> {
    let raw: Vec<SpellJson> = serde_json::from_str(json)?;
    Ok(raw.into_iter().filter_map(spell_from_json).collect())
}

fn spell_from_json(raw: SpellJson) -> Option<Spell> {
    let id = raw.id.filter(|s| !s.is_empty()).or_else(|| {
        eprintln!("WARNING: skipping spell with missing id");
        None
    })?;
    let source = raw.source.or_else(|| {
        eprintln!("WARNING: skipping spell '{id}': missing source");
        None
    })?;
    let level = raw.level.or_else(|| {
        eprintln!("WARNING: skipping spell '{id}': missing level");
        None
    })?;
    let school = raw.school.or_else(|| {
        eprintln!("WARNING: skipping spell '{id}': missing school");
        None
    })?;
    let concentration = raw.concentration.or_else(|| {
        eprintln!("WARNING: skipping spell '{id}': missing concentration");
        None
    })?;
    let ritual = raw.ritual.or_else(|| {
        eprintln!("WARNING: skipping spell '{id}': missing ritual");
        None
    })?;
    let raw_components = raw.components.or_else(|| {
        eprintln!("WARNING: skipping spell '{id}': missing components");
        None
    })?;
    let available_classes = raw.available_classes.or_else(|| {
        eprintln!("WARNING: skipping spell '{id}': missing availableClasses");
        None
    })?;
    let translations = raw.translations.filter(|t| !t.is_empty()).or_else(|| {
        eprintln!("WARNING: skipping spell '{id}': missing translations");
        None
    })?;

    Some(Spell {
        id,
        source,
        level,
        school,
        concentration,
        ritual,
        components: SpellComponents {
            verbal: raw_components.verbal,
            somatic: raw_components.somatic,
            material: raw_components.material,
        },
        available_classes,
        translations,
    })
}

fn parse_monsters(json: &str) -> Result<Vec<Monster>, Box<dyn std::error::Error>> {
    let raw: Vec<MonsterJson> = serde_json::from_str(json)?;
    Ok(raw.into_iter().filter_map(monster_from_json).collect())
}

fn monster_from_json(raw: MonsterJson) -> Option<Monster> {
    let id = raw.id.filter(|s| !s.is_empty()).or_else(|| {
        eprintln!("WARNING: skipping monster with missing id");
        None
    })?;
    let source = raw.source.or_else(|| {
        eprintln!("WARNING: skipping monster '{id}': missing source");
        None
    })?;
    let monster_type = raw.monster_type.or_else(|| {
        eprintln!("WARNING: skipping monster '{id}': missing type");
        None
    })?;
    let size = raw.size.or_else(|| {
        eprintln!("WARNING: skipping monster '{id}': missing size");
        None
    })?;
    let alignment = raw.alignment.or_else(|| {
        eprintln!("WARNING: skipping monster '{id}': missing alignment");
        None
    })?;
    let api_abilities = raw.abilities.or_else(|| {
        eprintln!("WARNING: skipping monster '{id}': missing abilities");
        None
    })?;
    let raw_translations = raw.translations.filter(|t| !t.is_empty()).or_else(|| {
        eprintln!("WARNING: skipping monster '{id}': missing translations");
        None
    })?;

    let translations: HashMap<String, Translation> = raw_translations
        .into_iter()
        .filter_map(|(locale, t)| {
            let name = t.name.or_else(|| {
                eprintln!("WARNING: monster '{id}' locale '{locale}': missing name");
                None
            })?;
            let description = t.description.or_else(|| {
                eprintln!("WARNING: monster '{id}' locale '{locale}': missing description");
                None
            })?;
            let senses = t.senses.or_else(|| {
                eprintln!("WARNING: monster '{id}' locale '{locale}': missing senses");
                None
            })?;
            Some((locale, Translation {
                name,
                subtype: t.subtype,
                description,
                senses,
                languages: t.languages.unwrap_or_default(),
            }))
        })
        .collect();

    if translations.is_empty() {
        eprintln!("WARNING: skipping monster '{id}': no valid translations");
        return None;
    }

    Some(Monster {
        id,
        source,
        monster_type,
        size,
        alignment,
        challenge_rating: raw.challenge_rating.unwrap_or(0.0),
        armor_class: raw.armor_class.unwrap_or(10),
        max_hit_points: raw.max_hit_points.unwrap_or(0),
        hit_dice: raw.hit_dice.unwrap_or_default(),
        abilities: Abilities {
            str: ability_from_json(api_abilities.str),
            dex: ability_from_json(api_abilities.dex),
            con: ability_from_json(api_abilities.con),
            int: ability_from_json(api_abilities.int),
            wis: ability_from_json(api_abilities.wis),
            cha: ability_from_json(api_abilities.cha),
        },
        speeds: speeds_from_json(raw.speeds),
        skills: raw.skills,
        damage_affinities: raw.damage_affinities,
        condition_immunities: raw.condition_immunities,
        translations,
    })
}

fn speeds_from_json(raw: Option<SpeedsJson>) -> Speeds {
    match raw {
        None => Speeds { walk: None, fly: None, swim: None, climb: None, burrow: None, hover: false },
        Some(s) => Speeds {
            walk: s.walk,
            fly: s.fly,
            swim: s.swim,
            climb: s.climb,
            burrow: s.burrow,
            hover: s.hover.unwrap_or(false),
        },
    }
}

fn ability_from_json(raw: Option<AbilityJson>) -> Ability {
    match raw {
        Some(a) => Ability {
            value: a.value.unwrap_or(10),
            saving_throw_proficiency: a.saving_throw_proficiency,
        },
        None => Ability { value: 10, saving_throw_proficiency: None },
    }
}

fn parse_magical_items(json: &str) -> Result<Vec<MagicalItem>, Box<dyn std::error::Error>> {
    let raw: Vec<MagicalItemJson> = serde_json::from_str(json)?;
    Ok(raw.into_iter().filter_map(magical_item_from_json).collect())
}

fn magical_item_from_json(raw: MagicalItemJson) -> Option<MagicalItem> {
    let id = raw.id.filter(|s| !s.is_empty()).or_else(|| {
        eprintln!("WARNING: skipping magical item with missing id");
        None
    })?;
    let source = raw.source.or_else(|| {
        eprintln!("WARNING: skipping magical item '{id}': missing source");
        None
    })?;
    let item_type = raw.item_type.or_else(|| {
        eprintln!("WARNING: skipping magical item '{id}': missing type");
        None
    })?;
    let rarity = raw.rarity.or_else(|| {
        eprintln!("WARNING: skipping magical item '{id}': missing rarity");
        None
    })?;
    let raw_translations = raw.translations.filter(|t| !t.is_empty()).or_else(|| {
        eprintln!("WARNING: skipping magical item '{id}': missing translations");
        None
    })?;

    let translations = raw_translations
        .into_iter()
        .filter_map(|(locale, t)| {
            let name = t.name.or_else(|| {
                eprintln!("WARNING: magical item '{id}' locale '{locale}': missing name");
                None
            })?;
            let description = t.description.or_else(|| {
                eprintln!("WARNING: magical item '{id}' locale '{locale}': missing description");
                None
            })?;
            Some((locale, MagicalItemTranslation { name, subtype: t.subtype, description }))
        })
        .collect::<HashMap<_, _>>();

    if translations.is_empty() {
        eprintln!("WARNING: skipping magical item '{id}': no valid translations");
        return None;
    }

    Some(MagicalItem {
        id,
        source,
        item_type,
        rarity,
        attunement: raw.attunement.unwrap_or(false),
        translations,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn monster_from_json_valid() {
        let raw = MonsterJson {
            id: Some("goblin".into()),
            source: Some("mm2024".into()),
            monster_type: Some("humanoid".into()),
            size: Some("small".into()),
            alignment: Some("neutral_evil".into()),
            challenge_rating: Some(0.25),
            armor_class: Some(15),
            max_hit_points: Some(7),
            hit_dice: Some("2d6".into()),
            abilities: Some(AbilitiesJson {
                str: Some(AbilityJson { value: Some(8), saving_throw_proficiency: None }),
                dex: Some(AbilityJson { value: Some(14), saving_throw_proficiency: None }),
                con: Some(AbilityJson { value: Some(10), saving_throw_proficiency: None }),
                int: Some(AbilityJson { value: Some(10), saving_throw_proficiency: None }),
                wis: Some(AbilityJson { value: Some(8), saving_throw_proficiency: None }),
                cha: Some(AbilityJson { value: Some(8), saving_throw_proficiency: None }),
            }),
            skills: HashMap::new(),
            damage_affinities: HashMap::new(),
            condition_immunities: HashMap::new(),
            translations: Some(HashMap::from([(
                "en".into(),
                TranslationJson {
                    name: Some("Goblin".into()),
                    subtype: Some("Goblinoid".into()),
                    description: Some("A small creature.".into()),
                    speed: Some("30 ft.".into()),
                    senses: Some("Darkvision 60 ft.".into()),
                    languages: Some(vec!["Common".into(), "Goblin".into()]),
                },
            )])),
        };
        let monster = monster_from_json(raw).expect("should parse");
        assert_eq!(monster.id, "goblin");
        assert_eq!(monster.monster_type, "humanoid");
        assert_eq!(monster.abilities.str.value, 8);
        assert_eq!(monster.translations["en"].name, "Goblin");
        assert_eq!(monster.translations["en"].subtype, Some("Goblinoid".into()));
    }

    #[test]
    fn monster_from_json_missing_id_returns_none() {
        let raw = MonsterJson {
            id: None,
            source: None,
            monster_type: None,
            size: None,
            alignment: None,
            challenge_rating: None,
            armor_class: None,
            max_hit_points: None,
            hit_dice: None,
            abilities: None,
            skills: HashMap::new(),
            damage_affinities: HashMap::new(),
            condition_immunities: HashMap::new(),
            translations: None,
        };
        assert!(monster_from_json(raw).is_none());
    }

    #[test]
    fn ability_from_json_defaults_to_10_when_none() {
        let ability = ability_from_json(None);
        assert_eq!(ability.value, 10);
        assert!(ability.saving_throw_proficiency.is_none());
    }
}
